components/speedreader/rust/lib/src/readability/src/extractor.rs

```diff
@@ -49,18 +76,178 @@ where
+                break;
+            }
+        }
+    }
+
+    let mut meta_tags = Meta::default();
+    for node in meta_nodes.clone().into_iter() {
+        let data = node.as_element().unwrap();
+        if data.name.local != local_name!("meta") {
+            continue;
+        }
+        if let Some(property) = data.attributes.borrow().get(local_name!("property")) {
+            if let Some(ref content) = data.attributes.borrow().get(local_name!("content")) {
+                match property {
+                    "dc:title"
+                    | "dcterm:title"
+                    | "og:title"
+                    | "weibo:article:title"
+                    | "weibo:webpage:title"
+                    | "title"
+                    | "twitter:title" => {
+                        meta_tags.title = content.to_string();
+                    }
+                    "description"
+                    | "dc:description"
+                    | "dcterm:description"
+                    | "og:description"
+                    | "weibo:article:description"
+                    | "weibo:webpage:description"
+                    | "twitter:description" => {
+                        meta_tags.description = Some(
+                            content
+                                .find(". ")
+                                .map(|pos| &content[..pos])
+                                .unwrap_or(content)
+                                .to_string(),
+                        );
+                    }
+                    "dc:creator" | "dcterm:creator" | "author" => {
+                        meta_tags.author = Some(content.to_string());
+                    }
+                    _ => (),
+                }
+            }
+        } else if let Some(charset) = data.attributes.borrow().get(local_name!("charset")) {
+            meta_tags.charset = Some(charset.to_string());
+        } else if data
+            .attributes
+            .borrow()
+            .get(local_name!("http-equiv"))
+            .map(|e| e.to_ascii_lowercase() == "content-type")
+            .unwrap_or(false)
+        {
+            if let Some(content) = data.attributes.borrow().get(local_name!("content")) {
+                if let Some(charset) = content.split("charset=").nth(1) {
+                    meta_tags.charset = Some(charset.trim().to_string());
+                }
+            }
+        }
+    }
+
+    let mut meta = meta_jsonld.merge(meta_tags);
+
+    if meta.title.is_empty() {
+        if let Some(handle) = meta_nodes
+            .into_iter()
+            .find(|d| d.as_element().unwrap().name.local == local_name!("title"))
+        {
+            dom::extract_text(&handle, &mut meta.title, true);
+        }
+    }
+
+    // Clean title of html encoded attributes
+    if !meta.title.is_empty() {
+        if let Ok(s) = decode_html(&meta.title) {
+            meta.title = DECODED_HTML_TAGS.replace_all(&s, "").into();
```

Comment:
Instead of these bespoke / hacky ways of decoding HTML, could you use the "proper" equivilents in html5ever, that'll handle trickier cases?